### PR TITLE
[expo-location][expo-permissions] Refactor location permissions

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

A follow up to the https://github.com/expo/expo/pull/11565.

# How

Split location permissions into `foreground` and `background` permission.
Unfortunately, not all versions of Android or iOS have `background` permission, but for those, we can just ask for all permission.

If Android supports background permissions:
- `Permissions.LOCATION` is the same as `Permissions.LOCATION_FOREGROUND`
- `Permissions.LOCATION_FOREGROUND` only asks for `ACCESS_FINE_LOCATION` and `ACCESS_COARSE_LOCATION` 
- `Permissions.LOCATION_BACKGROUND` only asks for `ACCESS_BACKGROUND_LOCATION` 

If iOS supports background permissions:
- `Permissions.LOCATION` is the same as `Permissions.LOCATION_FOREGROUND`
- `Permissions.LOCATION_FOREGROUND` only asks for  `when in use` permission.
-  `Permissions.LOCATION_BACKGROUND` only ask for `always` permission.

> On Web, both `FOREGROUND` and `BACKGROUND` are handled by the same logic. Both of these permissions will just ask the location of the browser.

# Test Plan

(Tested on iOS 12, 13, 14, and Android 28, 29, 30)
- using NCL ✅
- using simple app ✅
```tsc
import * as React from 'react';
import { View, StyleSheet, Button } from 'react-native';

import * as Constants from 'expo-constants';
import * as MediaLibrary from 'expo-media-library';
import * as FileSystem from 'expo-file-system';
import * as Permissions from 'expo-permissions';

export default function App() {
  return (
    <View style={styles.container}>
      <Button
        title="get legacy"
        onPress={async () => {
          const permissions = await Permissions.getAsync(Permissions.LOCATION);
          console.log({ permissions });
        }}
      />
      <Button
        title="get foreground"
        onPress={async () => {
          const permissions = await Permissions.getAsync(Permissions.LOCATION_FOREGROUND);
          console.log({ permissions });
        }}
      />
      <Button
        title="get background"
        onPress={async () => {
          const permissions = await Permissions.getAsync(Permissions.LOCATION_BACKGROUND);
          console.log({ permissions });
        }}
      />
      <Button
        title="ask legacy"
        onPress={async () => {
          const permissions = await Permissions.askAsync(Permissions.LOCATION);
          console.log({ permissions });
        }}
      />
      <Button
        title="ask foreground"
        onPress={async () => {
          const permissions = await Permissions.askAsync(Permissions.LOCATION_FOREGROUND);
          console.log({ permissions });
        }}
      />
      <Button
        title="ask background"
        onPress={async () => {
          const permissions = await Permissions.askAsync(Permissions.LOCATION_BACKGROUND);
          console.log({ permissions });
        }}
      />
    </View>
  );
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingTop: Constants.statusBarHeight,
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  paragraph: {
    margin: 24,
    fontSize: 18,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  videoContainer: {
    alignSelf: 'center',
    height: 250,
    width: 250,
    backgroundColor: 'red',
  },
});
```

### CODE REVIEW

File: packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java [hunk 1]
```diff
@@ -148,35 +150,111 @@ public void onCreate(ModuleRegistry moduleRegistry) {
...
+  @ExpoMethod
+  public void getForegroundPermissionsAsync(final Promise promise) {
     if (mPermissionsManager == null) {
       promise.reject("E_NO_PERMISSIONS", "Permissions module is null. Are you sure all the installed Expo modules are properly linked?");
       return;
     }
     mPermissionsManager.getPermissions(result -> {
-      promise.resolve(handleLocationPermissions(result));
+      promise.resolve(handleForegroundLocationPermissions(result));
     }, Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION);
```
Comment (byCedric) [inline packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java:227]: Maybe it's a good idea to make the combination between `ACCESS_FINE_LOCATION` and `ACCESS_COARSE_LOCATION` optional. So devs could use either one too (without defining that in their manifest).

That should cover the usecase of:
- Exclude background location permission
- On android versions before the `ACCESS_BACKGROUND_LOCATION` permission

**Review by byCedric — APPROVED** [review]

File: packages/expo-location/ios/EXLocation/Requesters/EXBackgroundLocationPermissionRequester.m [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
++ (void)load
+{
+  alwaysAuthorizationSelector = NSSelectorFromString([@"request" stringByAppendingString:@"AlwaysAuthorization"]);
+}
+
+- (void)requestLocationPermissions
+{
+  if ([EXBaseLocationRequester isConfiguredForAlwaysAuthorization] && [self.locMgr respondsToSelector:alwaysAuthorizationSelector]) {
+    _wasAsked = true;
+    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleAppBecomingActive) name:UIApplicationDidBecomeActiveNotification object:nil];
```
Comment (bbarthec) [inline packages/expo-location/ios/EXLocation/Requesters/EXBackgroundLocationPermissionRequester.m:40]:
```suggestion
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(handleAppBecomingActive)
                                                 name:UIApplicationDidBecomeActiveNotification
                                               object:nil];
```

File: packages/expo-location/ios/EXLocation/Requesters/EXForegroundPermissionRequester.m [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+@implementation EXForegroundPermissionRequester
+
++ (NSString *)permissionType
+{
+  return @"locationForeground";
+}
+
++ (void)load
+{
+  whenInUseAuthorizationSelector = NSSelectorFromString([@"request" stringByAppendingString:@"WhenInUseAuthorization"]);
```
Comment (bbarthec) [inline packages/expo-location/ios/EXLocation/Requesters/EXForegroundPermissionRequester.m:20]:
```suggestion
  whenInUseAuthorizationSelector = NSSelectorFromString(@"requestWhenInUseAuthorization");
```

File: packages/expo-location/ios/EXLocation/Requesters/EXForegroundPermissionRequester.m [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+- (void)requestLocationPermissions
+{
+  if ([EXBaseLocationRequester isConfiguredForWhenInUseAuthorization] && [self.locMgr  respondsToSelector:whenInUseAuthorizationSelector]) {
```
Comment (bbarthec) [inline packages/expo-location/ios/EXLocation/Requesters/EXForegroundPermissionRequester.m:25]:
```suggestion
  if ([EXBaseLocationRequester isConfiguredForWhenInUseAuthorization] && [self.locationManager respondsToSelector:whenInUseAuthorizationSelector]) {
```

File: packages/expo-location/ios/EXLocation/Requesters/EXLocationPermissionRequester.m [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
++ (NSString *)permissionType
+{
+  return @"location";
+}
+
++ (void)load
+{
+  alwaysAuthorizationSelector = NSSelectorFromString([@"request" stringByAppendingString:@"AlwaysAuthorization"]);
+  whenInUseAuthorizationSelector = NSSelectorFromString([@"request" stringByAppendingString:@"WhenInUseAuthorization"]);
```
Comment (bbarthec) [inline packages/expo-location/ios/EXLocation/Requesters/EXLocationPermissionRequester.m:21]:
```suggestion
  alwaysAuthorizationSelector = NSSelectorFromString(@"requestAlwaysAuthorization");
  whenInUseAuthorizationSelector = NSSelectorFromString([@"requestWhenInUseAuthorization");
```

File: packages/expo-location/ios/EXLocation/Requesters/EXLocationPermissionRequester.m [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+- (void)requestLocationPermissions
+{
+  if ([EXBaseLocationRequester isConfiguredForAlwaysAuthorization] && [self.locMgr respondsToSelector:alwaysAuthorizationSelector]) {
+    ((void (*)(id, SEL))objc_msgSend)(self.locMgr, alwaysAuthorizationSelector);
+  } else if ([EXBaseLocationRequester isConfiguredForWhenInUseAuthorization] && [self.locMgr respondsToSelector:whenInUseAuthorizationSelector]) {
+    ((void (*)(id, SEL))objc_msgSend)(self.locMgr, whenInUseAuthorizationSelector);
```
Comment (bbarthec) [inline packages/expo-location/ios/EXLocation/Requesters/EXLocationPermissionRequester.m:29]:
```suggestion
  if ([EXBaseLocationRequester isConfiguredForAlwaysAuthorization] && [self.locationManager respondsToSelector:alwaysAuthorizationSelector]) {
    ((void (*)(id, SEL))objc_msgSend)(self.locationManager, alwaysAuthorizationSelector);
  } else if ([EXBaseLocationRequester isConfiguredForWhenInUseAuthorization] && [self.locationManager respondsToSelector:whenInUseAuthorizationSelector]) {
    ((void (*)(id, SEL))objc_msgSend)(self.locationManager, whenInUseAuthorizationSelector);
```

File: packages/expo-location/src/Location.ts [hunk 7]
```diff
@@ -157,16 +157,51 @@ export async function reverseGeocodeAsync(
  * Gets the current state of location permissions.
  */
 export async function getPermissionsAsync(): Promise<LocationPermissionResponse> {
```
Comment (bbarthec) [inline packages/expo-location/src/Location.ts:159]:
```suggestion
 * Gets the current state of location permissions.
 * @deprecated User `getForegroundPermissions` or `getBackgroundPermissions` instead.
 */
export async function getPermissionsAsync(): Promise<LocationPermissionResponse> {
```

File: packages/expo-location/src/Location.ts [hunk 8]
```diff
@@ -157,16 +157,51 @@ export async function reverseGeocodeAsync(
...
+  console.warn(
+    `"getPermissionsAsync" is now deprecated. Please, use "getForegroundPermissions" or "getBackgroundPermissions" instead.`
+  );
   return await ExpoLocation.getPermissionsAsync();
 }
 
 /**
  * Requests the user to grant location permissions.
  */
 export async function requestPermissionsAsync(): Promise<LocationPermissionResponse> {
```
Comment (bbarthec) [inline packages/expo-location/src/Location.ts:169]:
```suggestion
 * Requests the user to grant location permissions.
 * @deprecated Use `requestForegroundPermissionsAsync` or `requestBackgroundPermissionsAsync` instead.
 */
export async function requestPermissionsAsync(): Promise<LocationPermissionResponse> {
```

File: packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+      putString(PermissionsResponse.EXPIRES_KEY, PermissionsResponse.PERMISSION_EXPIRES_NEVER)
+      putBoolean(PermissionsResponse.CAN_ASK_AGAIN_KEY, backgroundLocation.canAskAgain)
+      putBoolean(PermissionsResponse.GRANTED_KEY, permissionsStatus == PermissionsStatus.GRANTED)
+    }
+  }
+
+  @RequiresApi(Build.VERSION_CODES.Q)
+  private fun parseAndroidPermissionsForAndroidQ(permissionsResponse: Map<String, PermissionsResponse>): Bundle {
+    return Bundle().apply {
+
```
Comment (bbarthec) [inline packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt:57]:
```suggestion
```

File: packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      val statuses = listOf(accessFineLocation.status, accessCoarseLocation.status, accessBackgroundLocation.status)
+
+      val status = when {
+        statuses.all { it == PermissionsStatus.GRANTED } -> {
+          PermissionsStatus.GRANTED
+        }
+        statuses.all { it == PermissionsStatus.DENIED } -> {
+          PermissionsStatus.DENIED
+        }
```
Comment (bbarthec) [inline packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt:72]:
```suggestion
```
I don't really see if we need this 🤔

**Review by bbarthec — APPROVED** [review]
LGTM 👍 
Left some small comments & suggestions 😉

File: packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java (hunk 1 above)
Comment (lukmccall) [reply packages/expo-location/android/src/main/java/expo/modules/location/LocationModule.java:227]: I think is a good idea but we should make those changes in a separate PR.

File: packages/expo-location/ios/EXLocation/Requesters/EXBackgroundLocationPermissionRequester.m [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+}
+
++ (NSString *)permissionType
+{
+  return @"locationBackground";
+}
+
++ (void)load
+{
+  alwaysAuthorizationSelector = NSSelectorFromString([@"request" stringByAppendingString:@"AlwaysAuthorization"]);
```
[the remark this answers is not in the document]
Comment (lukmccall) [reply packages/expo-location/ios/EXLocation/Requesters/EXBackgroundLocationPermissionRequester.m:33]: This is a trick to outsmart Apple's review and I think we should leave it.

**Pull request merged**